Execution API spec (#572)
luna/enso: doc/design/engine/engine-services.md

### Capabilities
The capability management features work with the following capabilities.

#### `text/canEdit`
This capability states that the capability has the ability to perform both

> `text/applyEdit` and `text/save` for the specified file.

Reviewer: I think it's worth creating a section per capability that states which messages are enabled and/or disabled by the capability. See below. 

```suggestion
This capability states that the client has the ability to make edits to a given
file.

##### Enables

- `text/applyEdit`
- `text/save`

##### Disables
None
```

Please do this (or something similar) for the remaining capabilities.

##### Errors
TBC

> ### Execution Management

Reviewer: This should have a brief blurb about what execution management means.
```suggestion
### Execution Management
The execution management portion of the language server API deals with exposing
fine-grained control over program and expression execution to the clients of
the language server. This is incredibly important for enabling the high levels
of interactivity required by Enso Studio.
```

> #### Types

Reviewer: ```suggestion
#### Types
The execution management API exposes a set of common types used by many of its 
messages.
```

##### `ExpressionId`
An identifier used for Enso expressions.

```typescript
type ExpressionId = UUID;
```

##### `ContextId`
An identifier used for execution contexts.

```typescript
type ContextId = UUID;
```

##### `StackItem`
A representation of an executable position in code, used by the execution APIs.

> `MethodCall` is a call performed at the top of the stack, to initialize the

Reviewer: I don't see `MethodCall` in the definition at all. Do you mean `ExplicitCall`?

substituted for the unambiguous singleton instance.

`LocalCall` is a call corresponding to "entering a function call".

```typescript
type StackItem = ExplicitCall | LocalCall

interface ExplicitCall {
  methodPointer: MethodPointer;
  thisArgumentExpression?: String;
  positionalArgumentsExpressions: String[];
}

interface LocalCall {
  expressionId: ExpressionId;
}
```

##### `MethodPointer`
Points to a method definition.

```typescript
{
  file: Path;
  definedOnType: String;
  name: String;
}
```

##### `ExpressionValueUpdate`

```typescript
{
  id: ExpressionId;
  type?: String;
  shortValue?: String;
  methodCall?: MethodPointer;
}
```

#### `executionContext/create`
Sent from the client to the server to create a new execution context.

##### Parameters
```typescript
{
  contextId: ContextId;
}
```

##### Result
```typescript
{
  canModify: CapabilityRegistration;
  receivesEvents: CapabilityRegistration;
}
```

##### Errors
No known errors.

#### `executionContext/destroy`
Sent from the client to the server destroy an execution context and free its
resources.

##### Parameters
```typescript
{
  contextId: ContextId;
}
```

##### Result
```typescript
null
```

##### Errors

> - [`AccessDeniedError`](#accessdeniederror) when the user does not hold the `executionContext/canModify` capability for this context.

Reviewer: Please wrap this line.

#### `executionContext/fork`
Sent from the client to the server to duplicate an execution context, creating
an independent copy, containing all the data precomputed in the first one.

##### Parameters
```typescript
{
  contextId: ContextId;
  newContextId: ContextId;
}
```

##### Result
```typescript
{
  canModify: CapabilityRegistration;
  receivesEvents: CapabilityRegistration;
}
```

##### Errors
No known errors.

#### `executionContext/push`
Sent from the client to the server move the execution context to a new location
deeper down the stack.

##### Parameters
```typescript
{
  contextId: ContextId;
  stackItem: StackItem;
}
```

##### Result
```typescript
null
```

##### Errors
- [`StackItemNotFoundError`](#stackitemnotfounderror) when the request stack item could not be found.

> - [`AccessDeniedError`](#accessdeniederror) when the user does not hold the `executionContext/canModify` capability for this context.

Reviewer: Line wrapping.

#### `executionContext/pop`
Sent from the client to the server move the execution context up the stack,
corresponding to the client clicking out of the current breadcrumb.

##### Parameters
```typescript
{
  contextId: ContextId;
}
```

##### Result
```typescript
null
```

##### Errors

> - [`AccessDeniedError`](#accessdeniederror) when the user does not hold the `executionContext/canModify` capability for this context.

Reviewer: Wrapping

```typescript
"error" : {
  "code" : 1000,

>   "message" : String

Author: yes
